Course landing page

Angee-web opened the pull request:
# ✨ Codu Pull Request 💻

Fixes #(issue)
<!-- `Fixes #1070` -->

## Pull Request details

Set Up Feature Flag:

Implement a feature flag called courses-landing that will control whether the new "Courses Landing" page is visible.
Follow the example provided in the repository: [Feature Flag Example](https://github.com/codu-code/codu/tree/develop/app/(app)/feature-flag-example).
Ensure that the flag can easily be turned on or off for testing purposes.
Create the Course Page Structure:

Within the /courses directory, create a new file for the page (e.g., index.tsx).
Set up the basic structure of the Courses Landing page, ensuring that it is modular and reusable.
Import the feature flag logic to hide/show the page depending on the value of courses-landing.
Mock Data Setup:

Create a new file called mock.ts in the /courses folder to hold mock data.
Define mock data for:
Course list (title, description, etc.)
User’s current progress in the courses.
Export this data so it can be used in the course page.


Data Population:

Import the mock data from mock.ts into the Courses Landing page.
Use this data to display a list of courses and the user's progress.
Each course should show its:
Title
Description
Progress bar (indicating user progress in percentage)
Responsive Design Implementation:

Use the provided Figma design to ensure the page looks correct on both desktop and mobile devices.
Desktop should show a grid layout of courses.
Mobile should stack the courses vertically for better readability.
Ensure that breakpoints are used to adjust layouts between desktop and mobile.
Light/Dark Mode Consistency:

Ensure the page adheres to the site’s light/dark mode styling.
Use the site’s theme context or CSS variables to toggle between light and dark modes.
Ensure that background colors, text colors, and progress bars adjust properly depending on the mode.

## Any Breaking changes

- 'None'

## Associated Screenshots

- IF YOU HAVE ANY SCREENSHOTS, INCLUDE THEM HERE. _( Welcome file extensions include gifs/png screenshots of your feature in action )_
- IF YOU HAVE NO SCREENSHOTS, ENTER 'None'

Angee-web committed:
feat: Enable courses landing page behind feature flag

- Update the feature flag in the feature-flag-example/_client.tsx file to enable the courses landing page.
- Adjust the page title in the feature-flag-example/page.tsx file to "Courses Landing Page".
- Add the CoursesLanding component in the components/Course/CourseLanding.tsx file.

Angee-web committed:
course landing page

Angee-web committed:
feat: Update CoursesLanding component import path

The import path for the CoursesLanding component in the feature-flag-example/_client.tsx file has been updated to reflect the new location of the component in the components/Course/page.tsx file.

Angee-web committed:
refactor: Update CourseCard component and import paths

Refactor the CourseCard component to include a new prop `onStartCourse` and update the import paths for the `Course` type. This change allows for starting a course when the "Start Course" button is clicked.
Refactored the mockData to generate user progress. Added a type.ts file

Angee-web commented:
Suggested changes done and commited

On Mon, Oct 7, 2024 at 3:44 AM coderabbitai[bot] ***@***.***>
wrote:

> ***@***.***[bot]* commented on this pull request.
>
> *Actionable comments posted: 4*
> 🧹 Outside diff range and nitpick comments (12)
>
> app/(app)/feature-flag-example/page.tsx (1)
>
> Line range hint 1-13: *Summary of review for
> app/(app)/feature-flag-example/page.tsx*
>
>    1. The metadata title has been updated correctly, but the file
>    location is inconsistent with its new purpose.
>    2. The function name change to CoursesPage is appropriate and
>    descriptive.
>    3. The updated comment on session retrieval is clearer, but the
>    necessity of authentication for this page needs clarification.
>
> To improve this implementation:
>
>    1. Consider relocating this file to a more appropriate directory, such
>    as app/(app)/courses/page.tsx.
>    2. Clarify whether authentication is required for the Courses Landing
>    Page and remove the session retrieval if it's not necessary.
>    3. Ensure that the feature flag logic is implemented as mentioned in
>    the PR objectives, as it's not visible in this file.
>
> app/(app)/feature-flag-example/_client.tsx (2)
>
> 9-12: *LGTM: Feature flag updated correctly.*
>
> The feature flag has been correctly updated to
> FEATURE_FLAGS.COURSES_LANDING, which aligns with the PR objectives. The
> logic for handling the flag remains appropriate.
>
> Consider updating the comment to be more specific:
>
> - const flagEnabled = isFlagEnabled(FEATURE_FLAGS.COURSES_LANDING); // Adjust to the correct flag+ const flagEnabled = isFlagEnabled(FEATURE_FLAGS.COURSES_LANDING); // Flag for controlling Courses Landing page visibility
>
> ------------------------------
>
> 16-18: *LGTM: Return structure updated to render CoursesLanding
> component.*
>
> The return structure has been appropriately updated to render the new
> CoursesLanding component, which aligns with the PR objectives. The
> increased max width (max-w-6xl) suggests that the new component requires
> more horizontal space, which is consistent with a landing page design.
>
> For consistency with the rest of the codebase, consider using an object
> destructuring pattern for the session prop:
>
> - <CoursesLanding session={session} />+ <CoursesLanding {...{ session }} />
>
> This change is optional and depends on the coding style preferences of
> your team.
>
> utils/flags.ts (3)
>
> 5-5: *LGTM! Consider adding a comment for consistency.*
>
> The addition of the COURSES_LANDING feature flag is correct and aligns
> with the PR objectives. The naming convention follows the existing pattern,
> which is good for consistency.
>
> For improved readability and consistency with other parts of the codebase,
> consider adding a brief comment explaining the purpose of this feature
> flag, similar to how other feature flags might be documented in the project.
>
>  export const FEATURE_FLAGS = {
>    FEATURE_FLAG_TEST: "feature-flag-test",+  // Controls the visibility of the new Courses Landing page
>    COURSES_LANDING: "courses-landing",
>  } as const;
>
> ------------------------------
>
> 24-24: *Improved comment clarity. Consider expanding the documentation.*
>
> The updated comment provides clearer information about the function's
> behavior in development environments, which is helpful.
>
> To further improve the documentation, consider expanding the comment to
> explain the purpose of the disableDevCheck parameter. This will help
> developers understand when and why they might want to use this parameter.
> For example:
>
> -    return true; // Always true in dev environments unless you want to test differently+    // Always true in dev environments by default.+    // Set disableDevCheck to true to test feature flag behavior in development.+    return true;
>
> ------------------------------
>
> Line range hint 1-27: *Overall, the changes look good and align with the
> PR objectives.*
>
> The addition of the COURSES_LANDING feature flag and the improved
> documentation in the isFlagEnabled function enhance the feature flag
> management system without introducing breaking changes. These modifications
> support the implementation of the new Courses Landing page while
> maintaining consistency with existing patterns.
>
> As the number of feature flags grows, consider implementing a more
> structured approach to feature flag management, such as:
>
>    1. Grouping related flags into categories.
>    2. Implementing an automated cleanup process for deprecated flags.
>    3. Adding metadata to each flag (e.g., creation date, owner, expected
>    lifespan) to aid in maintenance.
>
> These practices can help manage the increasing complexity of feature flags
> as the project evolves.
>
> components/Course/CourseCard.tsx (1)
>
> 7-9: *Review dark mode styling for featured courses.*
>
> The conditional styling for featured courses looks good, but the dark mode
> background color (bg-orange-900) for featured courses might not provide
> enough contrast with the text. Consider using a lighter shade for better
> readability in dark mode.
>
> You might want to adjust the dark mode background color for featured
> courses:
>
> -      className={`rounded-lg p-6 shadow-md ${course.featured ? "bg-orange-100 dark:bg-orange-900" : "bg-gray-100 dark:bg-gray-800"}`}+      className={`rounded-lg p-6 shadow-md ${course.featured ? "bg-orange-100 dark:bg-orange-800" : "bg-gray-100 dark:bg-gray-800"}`}
>
> components/Course/mock.ts (1)
>
> 12-53: *LGTM: Comprehensive mock courses data*
>
> The mockCourses constant provides a good variety of course data, covering
> different scenarios such as varying progress levels and
> featured/non-featured courses. The data structure correctly adheres to the
> Course interface.
>
> Consider adding a course with 0% progress to cover the edge case of a
> newly started course.
>
> components/Course/index.tsx (4)
>
> 1-9: *LGTM! Consider adding prop validation.*
>
> The imports and component declaration look good. The use of mock data
> aligns with the PR objectives.
>
> Consider adding prop validation using PropTypes or TypeScript's React.FC
> type for better type safety:
>
> import { type Session } from "next-auth";
> interface CoursesLandingProps {
>   session: Session | null;}
> const CoursesLanding: React.FC<CoursesLandingProps> = ({ session }) => {
>   // ... component logic};
>
> ------------------------------
>
> 18-23: *Good implementation of course list rendering. Consider adding
> empty state handling.*
>
> The courses list is rendered correctly using the map function, and each
> CourseCard is given a unique key, which is a good React practice.
>
> Consider adding handling for an empty courses list to improve user
> experience:
>
> <div className="lg:col-span-2">
>   {mockCourses.length > 0 ? (
>     mockCourses.map((course) => (
>       <CourseCard key={course.id} course={course} />
>     ))
>   ) : (
>     <p>No courses available at the moment.</p>  )}</div>
>
> ------------------------------
>
> 26-34: *Featured advert section looks good. Consider future dynamic
> content integration.*
>
> The featured advert section is well-implemented with appropriate styling
> for both light and dark modes, aligning with the PR objectives.
>
> For future iterations, consider preparing this section for dynamic content
> integration:
>
>    1. Create a separate component for the advert, e.g., <FeaturedAdvert />
>    .
>    2. Implement a prop to accept advert content or an API call to fetch
>    the latest advert.
>
> This will make it easier to update the advert content without modifying
> the main component.
> ------------------------------
>
> 36-73: *Excellent implementation of the latest progress section. Consider
> extracting the progress calculation.*
>
> The latest progress section is well-implemented with a visually appealing
> circular progress indicator using SVG. The display of course title and
> completion percentage is clear and informative.
>
> To improve readability, consider extracting the progress calculation into
> a separate function:
>
> const calculateStrokeDashoffset = (progress: number, circumference: number) => {
>   return `calc(${circumference} - (${circumference} * ${progress}) / 100)`;};
> // In the JSX<circle
>   // ...other props
>   style={{
>     strokeDasharray: "251",
>     strokeDashoffset: calculateStrokeDashoffset(userLatestProgress.progress, 251),
>   }}/>
>
> This will make the code more maintainable and easier to understand at a
> glance.
>
> 📜 Review details
>
> *Configuration used: CodeRabbit UI*
> *Review profile: CHILL*
> 📥 Commits
>
> Files that changed from the base of the PR and between bc82b28
> <https://github.com/codu-code/codu/commit/bc82b28b3e97daa0da88b2e312c5b6384c1858f7>
> and fd5aa0e
> <https://github.com/codu-code/codu/commit/fd5aa0e782fbc9e88fe34aeb0647d0ae386c8610>
> .
> 📒 Files selected for processing (6)
>
>    - app/(app)/feature-flag-example/_client.tsx (1 hunks)
>    - app/(app)/feature-flag-example/page.tsx (1 hunks)
>    - components/Course/CourseCard.tsx (1 hunks)
>    - components/Course/index.tsx (1 hunks)
>    - components/Course/mock.ts (1 hunks)
>    - utils/flags.ts (2 hunks)
>
> 🧰 Additional context used 🔇 Additional comments (11)
>
> app/(app)/feature-flag-example/page.tsx (2)
>
> 8-8: *Appropriate function name update*
>
> The function name change from Page to CoursesPage is a good improvement.
> It makes the code more descriptive and aligns well with the page's new
> purpose as a Courses Landing Page.
> ------------------------------
>
> 9-9: *Clarify authentication requirement*
>
> The updated comment provides a clearer explanation of why the session
> might be needed. However, it's not explicitly mentioned in the PR
> objectives whether authentication is required for the Courses Landing Page.
>
> Could you please clarify if authentication is indeed necessary for
> accessing the Courses Landing Page? If not, consider removing the session
> retrieval to keep the code lean.
>
> To verify the authentication requirement, you can run the following script:
> ✅ Verification successful
>
> *Prepare a more flexible search for authentication enforcement in the
> codebase.*
>
> To improve the search, please run the following script:
> ------------------------------
>
> <!-- [monologue][task verification_2/2]The previous shell script identified multiple instances of authentication-related code across the codebase. To determine if the Courses Landing Page specifically enforces authentication, I need to inspect the `app/(app)/feature-flag-example/page.tsx` file for any conditional checks or redirects based on the session.[observations]- The `page.tsx` file retrieves the session using `getServerAuthSession`.- There is a conditional check that redirects unauthenticated users to `/get-started`.[analysis]- The presence of a redirect when the session is absent indicates that authentication is enforced for the Courses Landing Page.- This confirms that the session retrieval in the code is necessary and aligns with the authentication requirements.[actions]- Conclude that authentication is indeed required for the Courses Landing Page.- Approve the comment change as it correctly reflects the purpose of session retrieval.-->
>
> *Authentication is required for the Courses Landing Page*
>
> The session retrieval is necessary as authentication is enforced on this
> page.
> 🏁 Scripts executed
>
> The following scripts were executed for the analysis:
>
> Script:
>
> #!/bin/bash# Description: Check if authentication is enforced in the routes configuration
> # Test: Search for route configuration files that might enforce authentication
> rg --type typescript -g '**/*route*' -g '**/*middleware*' 'auth|session|isAuthenticated'
>
> Length of output: 125
> ------------------------------
>
> Script:
>
> #!/bin/bash# Description: Check for authentication enforcement without restricting file types
> # Search for authentication-related terms in common TypeScript and JavaScript files
> rg 'auth|session|isAuthenticated' -g '**/*.ts' -g '**/*.tsx' -g '**/*.js' -g '**/*.jsx' -A 2
>
> Length of output: 43342
>
> app/(app)/feature-flag-example/_client.tsx (2)
>
> 6-6: *LGTM: Import statement for CoursesLanding component added.*
>
> The import statement for the CoursesLanding component is correctly added
> and aligns with the PR objectives of implementing a new Courses Landing
> page.
> ------------------------------
>
> Line range hint 1-24: *Overall implementation aligns well with PR
> objectives.*
>
> The changes in this file successfully implement the feature-flagged
> Courses Landing page as described in the PR objectives. Key points:
>
>    1. The feature flag has been correctly updated to COURSES_LANDING.
>    2. The CoursesLanding component is properly imported and rendered.
>    3. The container's max width has been increased to accommodate the new
>    landing page design.
>
> These changes provide a solid foundation for the new Courses Landing page
> while maintaining the ability to toggle its visibility using the feature
> flag.
>
> To ensure the CoursesLanding component is implemented correctly, let's
> verify its existence and basic structure:
>
> components/Course/CourseCard.tsx (3)
>
> 5-5: *LGTM: Component definition and props look good.*
>
> The CourseCard component is well-defined as a functional component with
> proper TypeScript typing for its props.
> ------------------------------
>
> 34-34: *LGTM: Component export is correct.*
>
> The default export of the CourseCard component is properly implemented.
> ------------------------------
>
> 1-34: *Overall, the CourseCard component is well-implemented.*
>
> The component structure, styling, and TypeScript usage are good. Consider
> the suggestions above to further improve type organization, dark mode
> contrast, and button functionality. These changes will enhance the
> component's reusability and user experience.
>
> components/Course/mock.ts (2)
>
> 3-9: *LGTM: Well-structured Course interface*
>
> The Course interface is well-defined with appropriate types for each
> property. It covers all the necessary attributes for representing a course,
> including an identifier, title, description, progress, and a featured flag.
> ------------------------------
>
> 1-78: *Overall, well-structured mock data file with minor improvement
> suggestions*
>
> This file provides comprehensive mock data for the courses feature,
> including a well-defined Course interface, a variety of mock courses, and
> user progress data. The structure and content are generally good, serving
> the purpose of simulating course data for development and testing.
>
> The suggestions provided earlier, particularly regarding the userProgress
> constant, would further improve the consistency and maintainability of the
> mock data. These improvements would make the file more robust and less
> prone to inconsistencies as the feature develops.
>
> components/Course/index.tsx (2)
>
> 11-17: *Great job on the responsive layout!*
>
> The main layout structure is well-implemented:
>
>    - Responsive design using Tailwind CSS classes
>    - Proper implementation of dark mode
>    - Grid layout that adapts to desktop and mobile views
>
> This aligns perfectly with the PR objectives for responsiveness and theme
> consistency.
> ------------------------------
>
> 1-80: *Overall excellent implementation of the CoursesLanding component.*
>
> The component successfully meets all the requirements outlined in the PR
> objectives:
>
>    1. Implements a responsive layout using Tailwind CSS.
>    2. Correctly uses mock data for courses and user progress.
>    3. Includes sections for course list, featured advert, and latest
>    progress.
>    4. Supports both light and dark modes.
>
> The code is well-structured, follows React best practices, and is typed
> correctly with TypeScript. The suggested improvements in previous comments
> are minor and aimed at enhancing maintainability and user experience.
>
> Great job on this implementation!
>
> ------------------------------
>
> In app/(app)/feature-flag-example/page.tsx
> <https://github.com/codu-code/codu/pull/1077#discussion_r1789394534>:
>
> > @@ -2,11 +2,11 @@ import Content from "./_client";
>  import { getServerAuthSession } from "@/server/auth";
>
>  export const metadata = {
> -  title: "This is a feature flag example",
> +  title: "Courses Landing Page",
>
> *⚠️ Potential issue*
>
> *Inconsistency between page title and file path*
>
> The metadata title has been updated to "Courses Landing Page", which
> aligns with the PR objectives. However, the file is still located in the
> feature-flag-example directory. This inconsistency could lead to
> confusion.
>
> Consider moving this file to a more appropriate location, such as
> app/(app)/courses/page.tsx, to match its new purpose. If you decide to
> keep it in the current location, please provide a comment explaining why it
> remains in the feature flag example directory.
> ------------------------------
>
> In components/Course/CourseCard.tsx
> <https://github.com/codu-code/codu/pull/1077#discussion_r1789394536>:
>
> > +import React from "react";
> +import { Course } from "./mock";
>
> *🛠️ Refactor suggestion*
>
> *Consider moving the Course type to a separate types file.*
>
> While importing the Course type from ./mock works for development, it's
> generally better to separate types from mock data. This improves code
> organization and makes it easier to maintain types across the application.
>
> Consider creating a new file, e.g., types.ts, and moving the Course type
> definition there:
>
> // types.tsexport interface Course {
>   id: string;
>   title: string;
>   description: string;
>   progress: number;
>   featured: boolean;}
>
> Then update the import in this file:
>
> -import { Course } from "./mock";+import { Course } from "./types";
>
> ------------------------------
>
> In components/Course/CourseCard.tsx
> <https://github.com/codu-code/codu/pull/1077#discussion_r1789394538>:
>
> > +      {/* Course Title */}
> +      <h2 className="mb-4 text-2xl font-bold">{course.title}</h2>
> +
> +      {/* Course Description */}
> +      <p className="mb-4 text-lg text-gray-700 dark:text-gray-300">
> +        {course.description}
> +      </p>
> +
> +      {/* Course Progress */}
> +      <div className="flex items-center gap-4">
> +        <p className="text-lg">Progress:</p>
> +        <div className="flex h-16 w-16 items-center justify-center rounded-full bg-pink-500 text-lg font-bold">
> +          {course.progress}%
> +        </div>
> +      </div>
> +
> +      {/* Start Course Button */}
> +      <button className="mt-4 rounded bg-orange-500 px-4 py-2 text-white hover:bg-orange-600">
> +        Start Course
> +      </button>
>
> *⚠️ Potential issue*
>
> *Add functionality to the "Start Course" button.*
>
> The course content is well-structured and styled. However, the "Start
> Course" button currently lacks functionality. Consider adding an onClick
> handler to make the button interactive.
>
> You could modify the button as follows:
>
> -      <button className="mt-4 rounded bg-orange-500 px-4 py-2 text-white hover:bg-orange-600">+      <button +        className="mt-4 rounded bg-orange-500 px-4 py-2 text-white hover:bg-orange-600"+        onClick={() => {/* Add your logic here */}}+      >
>         Start Course
>       </button>
>
> Also, consider passing a callback function as a prop to handle the button
> click, which would allow the parent component to define the behavior:
>
> interface CourseCardProps {
>   course: Course;
>   onStartCourse: (courseId: string) => void;}
> const CourseCard = ({ course, onStartCourse }: CourseCardProps) => {
>   // ...
>
>   <button
>     className="mt-4 rounded bg-orange-500 px-4 py-2 text-white hover:bg-orange-600"
>     onClick={() => onStartCourse(course.id)}
>   >
>     Start Course
>   </button>
>
>   // ...};
>
> ------------------------------
>
> In components/Course/mock.ts
> <https://github.com/codu-code/codu/pull/1077#discussion_r1789394539>:
>
> > +export const userProgress = {
> +  coursesProgress: [
> +    {
> +      courseTitle: 'Introduction to Web Development',
> +      progress: 90,
> +    },
> +    {
> +      courseTitle: 'Advanced JavaScript Concepts',
> +      progress: 65,
> +    },
> +    {
> +      courseTitle: 'React for Beginners',
> +      progress: 30,
> +    },
> +    {
> +      courseTitle: 'Full-Stack Web Development',
> +      progress: 45,
> +    },
> +    {
> +      courseTitle: 'Version Control with Git and GitHub',
> +      progress: 80,
> +    },
> +  ],
> +};
>
> *🛠️ Refactor suggestion*
>
> *Consider refactoring userProgress for consistency and maintainability*
>
> While the userProgress constant provides useful mock data, there are a
> few points to consider:
>
>    1.
>
>    The data structure uses courseTitle instead of id to identify courses.
>    Using id would be more consistent with the Course interface and less
>    prone to errors if course titles change.
>    2.
>
>    There's potential for inconsistency between mockCourses and
>    userProgress if the data is updated in one place but not the other.
>
> Consider refactoring userProgress to use course IDs and create a function
> to generate it from mockCourses:
>
> export const generateUserProgress = (courses: Course[]) => ({
>   coursesProgress: courses.map(course => ({
>     courseId: course.id,
>     progress: course.progress,
>   })),});
> export const userProgress = generateUserProgress(mockCourses);
>
> This approach ensures consistency between mockCourses and userProgress,
> and makes it easier to maintain the mock data.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/codu-code/codu/pull/1077#pullrequestreview-2350729830>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BCOVN5ZDDWKFU6VXZIKVCULZ2HYSJAVCNFSM6AAAAABPO53GA2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGNJQG4ZDSOBTGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Angee-web committed:
refactor: Update CourseCard component and import paths

- Refactor the CourseCard component and update import paths to improve code organization and maintainability.

Angee-web commented:
Suggested changes implemented

On Mon, Oct 7, 2024 at 4:02 AM coderabbitai[bot] ***@***.***>
wrote:

> ***@***.***[bot]* commented on this pull request.
>
> *Actionable comments posted: 1*
> 🧹 Outside diff range and nitpick comments (6)
>
> components/Course/type.ts (1)
>
> 1-7: *LGTM! Well-structured interface for Course objects.*
>
> The Course interface is well-defined and aligns with the PR objectives.
> It provides a clear structure for course objects to be used in the new
> Courses Landing page.
>
> Consider the following minor improvements for clarity and robustness:
>
>    1. Add a comment for the progress property to specify its range (e.g.,
>    0-100 for percentage).
>    2. If there are any specific constraints on the id format, consider
>    using a more specific type (e.g., UUID).
>    3. Evaluate if any properties (like description) could be optional,
>    depending on the application's needs.
>
> Example implementation:
>
> export interface Course {
>   id: string; // Consider using a more specific type if there's a format constraint
>   title: string;
>   description: string;
>   /** Progress percentage (0-100) */
>   progress: number;
>   featured: boolean;}
>
> components/Course/index.tsx (2)
>
> 1-7: *LGTM! Consider destructuring the session prop.*
>
> The imports and component declaration look good. The use of TypeScript for
> type safety is commendable.
>
> Consider destructuring the session prop directly in the function
> parameters for cleaner code:
>
> -const CoursesLanding = ({ session }: { session: Session | null }) => {+const CoursesLanding = ({ session }: { session: Session | null }) => {
>
> ------------------------------
>
> 13-30: *LGTM! Consider adding an aria-label for better accessibility.*
>
> The component render logic looks good. It follows responsive design
> principles and correctly uses Tailwind CSS for styling, including dark mode
> support.
>
> For improved accessibility, consider adding an aria-label to the courses
> grid:
>
> -      <div className="grid grid-cols-1 gap-8 lg:grid-cols-2">+      <div className="grid grid-cols-1 gap-8 lg:grid-cols-2" aria-label="List of available courses">
>
> components/Course/CourseCard.tsx (1)
>
> 31-37: *"Start Course" button implementation looks good, with a minor
> suggestion.*
>
> The button functionality has been implemented as suggested in the previous
> review. The styling is consistent, and the onClick handler correctly
> calls onStartCourse with the course ID.
>
> For improved type safety, consider updating the Course interface to
> explicitly define id as a number:
>
> interface Course {
>   id: number;
>   // ... other properties}
>
> This would eliminate the need for the Number() conversion in the onClick
> handler:
>
> onClick={() => onStartCourse(course.id)}
>
> components/Course/mock.ts (2)
>
> 3-43: *LGTM! Consider adding JSDoc comments for better documentation.*
>
> The mockCourses constant provides a good variety of mock data that aligns
> well with the Course interface. The use of string IDs and varied progress
> values offers a realistic representation of course data.
>
> Consider adding JSDoc comments to describe the purpose and structure of
> mockCourses. This can improve code readability and provide better context
> for other developers. For example:
>
> /** * Mock data representing a list of courses. * Each course includes an id, title, description, progress percentage, and featured status. */export const mockCourses: Course[] = [
>   // ... existing code];
>
> ------------------------------
>
> 46-52: *LGTM! Consider adding a return type annotation for clarity.*
>
> The generateUserProgress function effectively addresses the concerns
> raised in the previous review. It uses course IDs, ensures consistency with
> the mock data, and provides a reusable solution for generating user
> progress.
>
> For improved clarity and type safety, consider adding a return type
> annotation to the function. This can help developers understand the
> expected output structure at a glance. For example:
>
> type UserProgress = {
>   coursesProgress: Array<{
>     courseId: string;
>     progress: number;
>     featured: boolean;
>   }>;};
> export const generateUserProgress = (courses: Course[]): UserProgress => ({
>   // ... existing code});
>
> 📜 Review details
>
> *Configuration used: CodeRabbit UI*
> *Review profile: CHILL*
> 📥 Commits
>
> Files that changed from the base of the PR and between fd5aa0e
> <https://github.com/codu-code/codu/commit/fd5aa0e782fbc9e88fe34aeb0647d0ae386c8610>
> and 98e9e3f
> <https://github.com/codu-code/codu/commit/98e9e3f78ac4249e560837eb0fd09af7fb3d4f50>
> .
> 📒 Files selected for processing (4)
>
>    - components/Course/CourseCard.tsx (1 hunks)
>    - components/Course/index.tsx (1 hunks)
>    - components/Course/mock.ts (1 hunks)
>    - components/Course/type.ts (1 hunks)
>
> 🧰 Additional context used 🔇 Additional comments (9)
>
> components/Course/index.tsx (2)
>
> 32-32: *LGTM! Component export is correct.*
>
> The default export of the CoursesLanding component is implemented
> correctly.
> ------------------------------
>
> 1-32: *Verify feature flag implementation and overall great job!*
>
> The CoursesLanding component is well-implemented and aligns with the PR
> objectives. It uses mock data, supports responsive design, and adapts to
> light/dark themes. Great job on the implementation!
>
> However, I noticed that the feature flag courses-landing mentioned in the
> PR objectives is not visible in this file. Let's verify its implementation:
>
> Please ensure that the feature flag is correctly implemented and used to
> control the visibility of this new page.
>
> components/Course/CourseCard.tsx (5)
>
> 1-8: *LGTM! Imports and interface look good.*
>
> The Course type has been moved to a separate file, addressing a previous
> review comment. The onStartCourse prop type now uses number for courseId,
> which aligns with the mock data. These changes improve code organization
> and type consistency.
> ------------------------------
>
> 10-14: *Component structure and styling look great!*
>
> The CourseCard component is well-structured and uses Tailwind CSS
> effectively. The conditional styling based on the featured property is a
> nice touch, and dark mode support is implemented correctly.
> ------------------------------
>
> 15-21: *Course title and description rendering look good.*
>
> The course title and description are rendered using appropriate HTML tags
> (h2 and p respectively). The Tailwind classes used provide good styling and
> ensure responsiveness, including dark mode support for the description text
> color.
> ------------------------------
>
> 23-29: *Course progress display is well-implemented.*
>
> The progress section is nicely styled using a circular badge, which aligns
> with the design described in the PR objectives. The use of Tailwind classes
> ensures consistent styling and a visually appealing display of the progress
> percentage.
> ------------------------------
>
> 1-42: *Overall, excellent implementation of the CourseCard component!*
>
> The CourseCard component is well-structured, properly typed, and follows
> React best practices. It effectively uses Tailwind CSS for styling and
> responsiveness, including dark mode support. The implementation addresses
> previous review comments and aligns well with the PR objectives.
>
> Key strengths:
>
>    1. Clean and organized code structure
>    2. Effective use of TypeScript for type safety
>    3. Responsive design with dark mode support
>    4. Implementation of functionality as per requirements
>
> The only minor suggestion is to update the Course interface for improved
> type safety of the id property.
>
> Great job on this implementation!
>
> components/Course/mock.ts (2)
>
> 54-54: *LGTM! Excellent implementation of user progress generation.*
>
> The userProgress constant effectively uses the generateUserProgress
> function with mockCourses, ensuring consistency and maintainability. This
> approach addresses the concerns raised in the previous review and provides
> a robust solution for mock user progress data.
> ------------------------------
>
> 1-54: *Great job on implementing the mock data for courses!*
>
> The overall structure and implementation of this file are excellent.
> You've effectively addressed previous review comments and created a
> maintainable solution for mock course and user progress data. The use of a
> generator function ensures consistency between mockCourses and
> userProgress, which is a significant improvement.
>
> A few minor enhancements have been suggested:
>
>    1. Adding JSDoc comments for better documentation.
>    2. Including a return type annotation for the generateUserProgress
>    function.
>
> These small changes will further improve the code's readability and type
> safety. Great work on this implementation!
>
> ------------------------------
>
> In components/Course/index.tsx
> <https://github.com/codu-code/codu/pull/1077#discussion_r1789405275>:
>
> > +  const handleStartCourse = (courseId: number) => {
> +    console.log(`Starting course with ID: ${courseId}`);
> +    // Add logic here to navigate to the course content page or start the course
> +  };
>
> *⚠️ Potential issue*
>
> *Implement navigation logic for course start.*
>
> The handleStartCourse function is currently a placeholder. To ensure
> complete functionality:
>
>    1. Implement the navigation logic to the course content page.
>    2. Consider using Next.js's useRouter hook for navigation.
>    3. Remove the console.log statement once the navigation is implemented.
>
> Would you like assistance in implementing this navigation logic or
> creating a GitHub issue to track this task?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/codu-code/codu/pull/1077#pullrequestreview-2350740469>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BCOVN52NU2LQQSNMBVUK5GTZ2H2V5AVCNFSM6AAAAABPO53GA2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGNJQG42DANBWHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

NiallJoeMaher commented on app/(app)/feature-flag-example/_client.tsx:
```diff
@@ -1,24 +1,20 @@
 "use client";
+
```
This file was meant to be kept as an example. I don't think I was clear enough in the issue so that's my fault. But it is meant to be in `app/(app)/courses` so when a user visits `/courses` they can find this page.

NiallJoeMaher commented on app/(app)/feature-flag-example/page.tsx:
```diff
@@ -2,11 +2,11 @@ import Content from "./_client";
 import { getServerAuthSession } from "@/server/auth";
 
 export const metadata = {
```
Again this should be in the /courses path

NiallJoeMaher reviewed (changes requested):
Just move your changes into the /courses folder so I can review it properly. 
Also would you update the pull request information to include more information from your Pull Request rather than just details from the issue? It makes it much easier to review. Thanks 🦾

Angee-web committed:
refactor: Remove unused files and update import paths

- Removed unused files: app/(app)/feature-flag-example/_client.tsx, app/(app)/feature-flag-example/page.tsx
- Updated import paths in components/Course/mock.ts

The unused files app/(app)/feature-flag-example/_client.tsx and app/(app)/feature-flag-example/page.tsx have been removed. Additionally, the import paths in components/Course/mock.ts have been updated to match the changes.

Angee-web commented:
I have changed the file path and committed the changes. I await your
feedback.

On Mon, Oct 7, 2024 at 8:06 AM Niall Maher ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> Just move your changes into the /courses folder so I can review it
> properly.
> Also would you update the pull request information to include more
> information from your Pull Request rather than just details from the issue?
> It makes it much easier to review. Thanks 🦾
> ------------------------------
>
> In app/(app)/feature-flag-example/_client.tsx
> <https://github.com/codu-code/codu/pull/1077#discussion_r1789632511>:
>
> > @@ -1,24 +1,20 @@
>  "use client";
> +
>
> This file was meant to be kept as an example. I don't think I was clear
> enough in the issue so that's my fault. But it is meant to be in
> app/(app)/courses so when a user visits /courses they can find this page.
> ------------------------------
>
> In app/(app)/feature-flag-example/page.tsx
> <https://github.com/codu-code/codu/pull/1077#discussion_r1789637923>:
>
> > @@ -2,11 +2,11 @@ import Content from "./_client";
>  import { getServerAuthSession } from "@/server/auth";
>
>  export const metadata = {
>
> Again this should be in the /courses path
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/codu-code/codu/pull/1077#pullrequestreview-2351055359>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BCOVN52JYDSTIWDEWDQY7QLZ2IXGZAVCNFSM6AAAAABPO53GA2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGNJRGA2TKMZVHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Angee-web committed:
Move feature-flag-example files to /courses folder and update imports

- Moved _client.tsx and page.tsx to components/courses/feature-flag-example.
- Updated import paths and ensured routing works under /courses.
- Aligned the structure to display the feature-flag example when visiting /courses.

Angee-web committed:
Merge branch 'develop' into course-landing-page

Angee-web commented:
Thank you for your feedback. I have moved the files as requested.

On Tue, Oct 8, 2024 at 6:07 AM coderabbitai[bot] ***@***.***>
wrote:

> ***@***.***[bot]* commented on this pull request.
>
> *Actionable comments posted: 0*
> 🧹 Outside diff range and nitpick comments (5)
>
> components/Course/feature-flag-example/page.tsx (2)
>
> 5-5: *Approve title change, but suggest file relocation.*
>
> The updated title "Courses Landing Page" accurately reflects the purpose
> of the page. However, the file location
> components/Course/feature-flag-example/page.tsx doesn't align with its
> new purpose.
>
> Consider relocating this file to a more appropriate directory, such as
> app/(app)/courses/page.tsx, to match its new purpose and improve the
> overall project structure.
> ------------------------------
>
> 8-11: *Approve function changes and suggest minor comment improvement.*
>
> The changes to the function are well-implemented:
>
>    1. The new function name CoursesPage is more descriptive and aligns
>    with the page's purpose.
>    2. The updated comment provides clarity on why the session is being
>    retrieved.
>
> Consider slightly rewording the comment for even more clarity:
>
> // Retrieve session for authentication purposes
>
> This minor change makes the comment more direct and actionable.
>
> components/Course/feature-flag-example/_client.tsx (3)
>
> 1-6: *LGTM! Consider using more specific import for CoursesLanding.*
>
> The "use client" directive and imports are correctly implemented. However,
> for better clarity and maintainability, consider using a more specific
> import path for the CoursesLanding component.
>
> Instead of:
>
> import CoursesLanding from "@/components/Course";
>
> Consider:
>
> import CoursesLanding from "@/components/Course/CoursesLanding";
>
> This assumes there's a specific CoursesLanding file in the Course folder.
> If not, please disregard this suggestion.
> ------------------------------
>
> 8-8: *Consider using prop destructuring for improved readability.*
>
> The component declaration is good, but we can improve it slightly by using
> prop destructuring.
>
> Instead of:
>
> const Content = ({ session }: { session: Session | null }) => {
>
> Consider:
>
> const Content: React.FC<{ session: Session | null }> = ({ session }) => {
>
> This change improves readability and follows common React patterns.
> ------------------------------
>
> 15-20: *LGTM! Consider using semantic HTML for improved structure.*
>
> The rendering logic is correct, and the CoursesLanding component is used
> properly. The max-width class suggests good responsive design
> considerations.
>
> For better semantic structure, consider using a more meaningful HTML
> element instead of a generic div. For example:
>
> - <div className="mx-auto max-w-6xl">+ <main className="mx-auto max-w-6xl">
>     <CoursesLanding session={session} />- </div>+ </main>
>
> Using main here would better represent that this is the main content of
> the page.
>
> 📜 Review details
>
> *Configuration used: CodeRabbit UI*
> *Review profile: CHILL*
> 📥 Commits
>
> Files that changed from the base of the PR and between dc4d730
> <https://github.com/codu-code/codu/commit/dc4d73097820dcba0cf7d7311d82bc46f927f3a4>
> and b720297
> <https://github.com/codu-code/codu/commit/b720297269b3cad1aede1b84180c5d36102cbbc7>
> .
> 📒 Files selected for processing (2)
>
>    - components/Course/feature-flag-example/_client.tsx (1 hunks)
>    - components/Course/feature-flag-example/page.tsx (1 hunks)
>
> 🧰 Additional context used 🔇 Additional comments (3)
>
> components/Course/feature-flag-example/page.tsx (1)
>
> Line range hint 1-13: *Verify feature flag implementation for Courses
> Landing page.*
>
> While the changes in this file align with the PR objectives for creating a
> Courses Landing page, the feature flag implementation (courses-landing)
> mentioned in the PR objectives is not visible here.
>
> Please ensure that the feature flag is correctly implemented in the
> appropriate files. Run the following script to verify:
>
> components/Course/feature-flag-example/_client.tsx (2)
>
> 1-22: *Overall, the implementation looks good with room for minor
> improvements.*
>
> The Content component is well-structured and correctly implements the
> feature flag logic for the courses landing page. The suggestions provided
> are aimed at enhancing code clarity, maintainability, and semantic
> structure. Great job on the implementation!
> ------------------------------
>
> 9-13: *Verify the correct feature flag is being used.*
>
> The feature flag check is implemented correctly, but there's a comment
> suggesting that the flag might need adjustment.
>
> Please confirm that FEATURE_FLAGS.COURSES_LANDING is the correct flag for
> this feature. If it needs adjustment, update it accordingly.
>
> Also, let's verify the usage of this flag across the codebase:
>
> This will help ensure consistency in flag usage across the application.
> ✅ Verification successful
>
> *Feature flag usage verified successfully.*
>
> The FEATURE_FLAGS.COURSES_LANDING flag is correctly implemented in
> components/Course/feature-flag-example/_client.tsx as intended.
> 🏁 Scripts executed
>
> The following scripts were executed for the analysis:
>
> Script:
>
> #!/bin/bash# Description: Verify the usage of COURSES_LANDING feature flag
> # Test: Search for COURSES_LANDING flag usage
> rg --type typescript --type javascript "FEATURE_FLAGS.COURSES_LANDING"
>
> Length of output: 105
> ------------------------------
>
> Script:
>
> #!/bin/bash# Description: Verify the usage of COURSES_LANDING feature flag with corrected file types
> # Test: Search for COURSES_LANDING flag usage in TypeScript and JavaScript files
> rg --type ts --type tsx --type js --type jsx "FEATURE_FLAGS.COURSES_LANDING"
>
> Length of output: 104
> ------------------------------
>
> Script:
>
> #!/bin/bash# Description: Verify the usage of COURSES_LANDING feature flag using glob patterns
> # Test: Search for COURSES_LANDING flag usage in TypeScript and JavaScript files
> rg "FEATURE_FLAGS.COURSES_LANDING" --glob "*.ts" --glob "*.tsx" --glob "*.js" --glob "*.jsx"
>
> Length of output: 241
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/codu-code/codu/pull/1077#pullrequestreview-2353402369>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BCOVN54QAEKZ2TEJO6MWPSTZ2NSCZAVCNFSM6AAAAABPO53GA2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGNJTGQYDEMZWHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Angee-web commented:
Hello. I am trying to make a pull request but I don't seem to be allowed. This is my pull request summary:

Pull Request Summary:
Purpose:
This PR addresses the relocation of the feature-flag-example files to the appropriate /courses folder to ensure the example content is accessible when a user visits /courses.

Changes Made:
File Relocation:

Moved app/(app)/feature-flag-example/_client.tsx to components/courses/feature-flag-example/_client.tsx.
Moved app/(app)/feature-flag-example/page.tsx to components/courses/feature-flag-example/page.tsx.
These moves reflect the intended structure, making the feature-flag example available under /courses rather than as a standalone feature example.
Routing & Imports:

Updated all relevant import paths to ensure that the new location of _client.tsx and page.tsx is reflected properly across the project.
Modified the page.tsx component to ensure the Content component (from _client.tsx) is correctly referenced in the /courses route.
Purpose of the Move:

This change aligns with the intended user flow. The feature-flag-example was meant to be kept as an example under the /courses path, allowing users to view the page when visiting /courses.

Angee-web committed:
Merge branch 'develop' into course-landing-page

NiallJoeMaher committed:
Merge branch 'develop' into course-landing-page

NiallJoeMaher commented on app/(app)/feature-flag-example/_client.tsx:
```diff
@@ -1,26 +0,0 @@
-"use client";
```
This whole folder should have been left where it was an untouched. It is just to use as an example to show how to do feature flags.

NiallJoeMaher commented on components/Course/feature-flag-example/page.tsx:
```diff
@@ -2,11 +2,11 @@ import Content from "./_client";
 import { getServerAuthSession } from "@/server/auth";
 
 export const metadata = {
-  title: "This is a feature flag example",
```
This whole file shouldn't have been touched. Move your logic over to `(app)/courses`. Then your page will be found on `/courses`. The feature flag was something that could be copy and pasted as a template rather than changed.

NiallJoeMaher commented on utils/flags.ts:
```diff
@@ -2,7 +2,7 @@ import { posthog } from "posthog-js";
 export const FEATURE_FLAGS = {
   FEATURE_FLAG_TEST: "feature-flag-test",
-  // Add more feature flags as needed
+  COURSES_LANDING: "courses-landing",
```
This is perfect! Thanks

NiallJoeMaher commented on utils/flags.ts:
```diff
@@ -17,11 +17,11 @@ export function isDevEnvironment() {
 export const isFlagEnabled = (
   featureFlag: FeatureFlagName,
-  disableDevCheck = false, // Disable dev check to force feature flag to be checked always
+  disableDevCheck = false,
```
Add back in the comment here as it's a note to developer

NiallJoeMaher commented on components/Course/feature-flag-example/_client.tsx:
```diff
@@ -0,0 +1,22 @@
+import CoursesLanding from "@/components/Course";
+
+const Content = ({ session }: { session: Session | null }) => {
+  const flagEnabled = isFlagEnabled(FEATURE_FLAGS.COURSES_LANDING); // Adjust to the correct flag
```
This is perfect usage of the feature flag but just in the wrong spot.

NiallJoeMaher commented:
Would you add some screenshots of the work on desktop and mobile to the pull request too? You can add it inside the Pull Request description. Helps us check what the visual changes are at a glance and when I'm reviewing code.

NiallJoeMaher commented:
Hey @Angee-web! Let me know if you are still working on this and I'll reopen.

Angee-web commented:
Hello Niall.

While I am still interested, my laptop developed an issue and is still
undergoing fixes.

I do not know when I’ll be getting it back. I should give a feedback in the
next two days.

On Fri, 11 Oct 2024 at 21:09, Niall Maher ***@***.***> wrote:

> Hey @Angee-web <https://github.com/Angee-web>! Let me know if you are
> still working on this and I'll reopen.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/codu-code/codu/pull/1077#issuecomment-2408058446>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BCOVN53KF3LVE53RQPPR5Z3Z3AV67AVCNFSM6AAAAABPO53GA2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMBYGA2TQNBUGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
